File: sdk/experimental/rs/src/telemetry/install_id.rs
//! Installation ID reader/writer (T-0724).
//!
//! Mirrors `go/runtime/telemetry/installid.go`: 32 raw bytes from
//! `getrandom`, stored at
//! `<XDG_STATE_HOME>/kit/telemetry/installation_id` with 0600 perms +
//! 0700 parent dir. The public API returns the lowercase hex SHA-256
//! of those bytes so the on-disk format stays canonical across
//! polyglot SDKs while the hashed string is what flows through
//! events.
//!
//! Concurrency: this implementation handles intra-process and
//! cross-process races via a tmp-file + rename strategy. If two
//! processes race on first-run the rename loser falls back to reading
//! whatever the winner wrote.

use sha2::{Digest, Sha256};
use std::fmt::Write as _;
use std::fs;
use std::io;
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};

const INSTALL_ID_SIZE: usize = 32;
const FILE_PERM: u32 = 0o600;
const DIR_PERM: u32 = 0o700;

/// Resolve `<XDG_STATE_HOME>/kit/telemetry/installation_id`, honoring
/// the XDG_STATE_HOME env var with a `$HOME/.local/state` fallback.
pub fn install_id_path() -> PathBuf {
    let base = std::env::var_os("XDG_STATE_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            dirs::home_dir()
                .unwrap_or_default()
                .join(".local")
                .join("state")
        });
    base.join("kit").join("telemetry").join("installation_id")
}

/// Return the persisted installation ID as a 64-char lowercase hex
/// SHA-256 digest. Generates + persists 32 random bytes on first call.
/// Cross-process race: if two callers race on the first generation the
/// rename loser re-reads whatever the winner wrote.
pub fn get_install_id() -> io::Result<String> {
    let p = install_id_path();

    // Fast path: file already exists.
    if p.exists() {
        let data = fs::read(&p)?;
        if data.len() != INSTALL_ID_SIZE {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!(
                    "install_id: file has wrong size {} bytes, expected {}",
                    data.len(),
                    INSTALL_ID_SIZE
                ),
            ));
        }
        return Ok(hex_sha256(&data));
    }

    // Slow path: generate + persist atomically.
    ensure_parent(&p)?;

    let mut fresh = [0u8; INSTALL_ID_SIZE];
    getrandom::getrandom(&mut fresh)
        .map_err(|e| io::Error::other(format!("getrandom: {e}")))?;

    let tmp = p.with_extension("new");
    fs::write(&tmp, fresh)?;
    fs::set_permissions(&tmp, fs::Permissions::from_mode(FILE_PERM))?;

    match fs::rename(&tmp, &p) {
        Ok(()) => Ok(hex_sha256(&fresh)),
        Err(_) if p.exists() => {
            // Lost the race to another process; read what they wrote.
            let _ = fs::remove_file(&tmp);
            let data = fs::read(&p)?;
            if data.len() != INSTALL_ID_SIZE {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!(
                        "install_id: file has wrong size {} bytes, expected {}",
                        data.len(),
                        INSTALL_ID_SIZE
                    ),
                ));
            }
            Ok(hex_sha256(&data))
        }
        Err(e) => {
            let _ = fs::remove_file(&tmp);
            Err(e)
        }
    }
}

/// Atomically replace the persisted bytes with 32 fresh random bytes
/// and return the new hex. Used by the `kit consent reset` flow.
pub fn rotate() -> io::Result<String> {
    let p = install_id_path();
    ensure_parent(&p)?;

    let mut fresh = [0u8; INSTALL_ID_SIZE];
    getrandom::getrandom(&mut fresh)
        .map_err(|e| io::Error::other(format!("getrandom: {e}")))?;

    let tmp = p.with_extension("new");
    // Clear any stale .new from a previous crashed writer before
    // re-creating it.
    let _ = fs::remove_file(&tmp);
    fs::write(&tmp, fresh)?;
    fs::set_permissions(&tmp, fs::Permissions::from_mode(FILE_PERM))?;
    fs::rename(&tmp, &p)?;
    Ok(hex_sha256(&fresh))
}

/// Create the parent directory of `path` with 0700 perms (idempotent).
/// Force-tightens existing dirs in case some other tool created the
/// parent at 0755.
fn ensure_parent(path: &Path) -> io::Result<()> {
    let parent = path.parent().ok_or_else(|| {
        io::Error::new(
            io::ErrorKind::InvalidInput,
            "install_id: path has no parent",
        )
    })?;
    fs::create_dir_all(parent)?;
    fs::set_permissions(parent, fs::Permissions::from_mode(DIR_PERM))?;
    Ok(())
}

fn hex_sha256(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    let result = h.finalize();
    let mut s = String::with_capacity(64);
    for b in result {
        // Infallible on String.
        let _ = write!(s, "{:02x}", b);
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hex_sha256_known_vector() {
        // SHA-256("") = e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
        let h = hex_sha256(&[]);
        assert_eq!(
            h,
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );
        assert_eq!(h.len(), 64);
    }
}

File: sdk/experimental/php/src/Telemetry/InstallId.php
<?php

declare(strict_types=1);

namespace HopTop\Kit\Telemetry;

use RuntimeException;

/**
 * Read / write the persistent installation_id used to attribute
 * pseudonymous telemetry events.
 *
 * Wire format mirrors the Go ground truth in
 * go/runtime/telemetry/installid.go: 32 raw bytes on disk, returned to
 * callers as the lowercase SHA-256 hex digest of those bytes. Storing
 * the raw bytes (not the hash) lets us salt or rotate downstream
 * without touching the file format.
 */
final class InstallId
{
    /**
     * Canonical on-disk path:
     *   $XDG_STATE_HOME/kit/telemetry/installation_id
     * (defaults to $HOME/.local/state).
     */
    public static function path(): string
    {
        $stateHome = getenv('XDG_STATE_HOME');
        if ($stateHome === false || $stateHome === '') {
            $home = (string) ($_SERVER['HOME'] ?? getenv('HOME') ?: '');
            $stateHome = $home . '/.local/state';
        }

        return $stateHome . '/kit/telemetry/installation_id';
    }

    /**
     * Returns the SHA-256 hex digest of the 32 raw bytes persisted at
     * the canonical path. Generates + persists on first call.
     *
     * Race-safe across concurrent first calls: a losing writer accepts
     * the bytes written by the winner.
     *
     * @throws RuntimeException If the on-disk file has the wrong size
     *         or rename to the canonical path fails irrecoverably.
     */
    public static function get(): string
    {
        $p = self::path();

        if (file_exists($p)) {
            return self::hashOrThrow($p);
        }

        $parent = dirname($p);
        if (!is_dir($parent) && !@mkdir($parent, 0o700, true) && !is_dir($parent)) {
            throw new RuntimeException("install_id: failed to mkdir {$parent}");
        }

        $fresh = random_bytes(32);
        $tmp = $p . '.new';
        if (file_put_contents($tmp, $fresh) !== 32) {
            throw new RuntimeException("install_id: failed to write tmp file {$tmp}");
        }
        @chmod($tmp, 0o600);

        // Atomic rename; if another process won the race, accept their bytes.
        if (!@rename($tmp, $p)) {
            if (file_exists($p)) {
                @unlink($tmp);

                return self::hashOrThrow($p);
            }
            @unlink($tmp);
            throw new RuntimeException("install_id: rename failed for {$p}");
        }

        return hash('sha256', $fresh);
    }

    /**
     * Rotate the installation_id: overwrite with fresh bytes and
     * return the new SHA-256 hex digest. Used by the consent flow
     * when the user revokes + re-grants.
     *
     * @throws RuntimeException If the rename to the canonical path
     *         fails.
     */
    public static function rotate(): string
    {
        $p = self::path();
        $parent = dirname($p);
        if (!is_dir($parent) && !@mkdir($parent, 0o700, true) && !is_dir($parent)) {
            throw new RuntimeException("install_id: failed to mkdir {$parent}");
        }

        $fresh = random_bytes(32);
        $tmp = $p . '.new';
        if (file_put_contents($tmp, $fresh) !== 32) {
            throw new RuntimeException("install_id: failed to write tmp file {$tmp}");
        }
        @chmod($tmp, 0o600);

        if (!@rename($tmp, $p)) {
            @unlink($tmp);
            throw new RuntimeException("install_id: rotate rename failed for {$p}");
        }

        return hash('sha256', $fresh);
    }

    /**
     * Test helper: remove the canonical file so the next get() call
     * regenerates. No-op when absent.
     */
    public static function resetForTest(): void
    {
        $p = self::path();
        if (file_exists($p)) {
            @unlink($p);
        }
    }

    /**
     * @throws RuntimeException If the file is the wrong size.
     */
    private static function hashOrThrow(string $p): string
    {
        $data = file_get_contents($p);
        if ($data === false) {
            throw new RuntimeException("install_id: failed to read {$p}");
        }
        $len = strlen($data);
        if ($len !== 32) {
            throw new RuntimeException(
                "install_id: file has wrong size {$len} bytes, expected 32"
            );
        }

        return hash('sha256', $data);
    }
}
